tests/linearizability/README.md

```diff
@@ -0,0 +1,102 @@
+# etcd Linearizability Testing
+
+Purpose of etcd linearizability tests is to validate that etcd upholds
+[API guarantees] even while experiencing failures common for a distributed system.
+
+This is done by comparing how real etcd cluster behaves versus a simplified model of etcd.
+During the test we continuously inject different types of failures and validate etcd correctness.
+
+[API guarantees]: https://etcd.io/docs/v3.5/learning/api_guarantees/
+
+## Motivation
+
+As proposed in [v3.5 data inconsistency postmortem], etcd needed a new framework
+for release qualification. A framework that can not only be able to prevent
+historical issues, but also discover new issues. At the time etcd had a big
+problem of test flakiness, so it was also important to easily distinguish between
+etcd problem and a bug in testing.
+
+In the end we were inspired by [Jepsen] and implemented a similar set of tests
+that would validate etcd [linearizability]. The framework is able to reproduce
+historical inconsistency issues by inject specific failpoints into etcd server.
+Flakes can be easily distinguished as verification is done on history of
+successful requests and does not depend on failure of any specific request.
+
+[v3.5 data inconsistency postmortem]: https://github.com/etcd-io/etcd/blob/main/Documentation/postmortems/v3.5-data-inconsistency.md#action-items
+[Jepsen]: https://github.com/jepsen-io/jepsen
+[linearizability]: https://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf
+
+
+## How it works
+
+Linearizability tests are not much different from standard e2e tests. They run
+e2e binaries to create a cluster and validate etcd behavior by sending requests
+to API. There are however to major differences. In addition to sending requests
+they also inject different failures into cluster (for example kill etcd process).
+Instead of running any specific test scenario (for example 1. put value 2. get
+value 3. compare if read equals put) they record all the history of requests
+and try to replay them against a simplified model of etcd.
+
+Steps in linearizability tests
+
+1. Setup a etcd cluster
+2. Send traffic to each member and record all requests
+3. Inject failures into the etcd clusters
+4. Validate request history against a etcd model
+
```

Comment:
It would be good to mentioned the relationship with to the 'functional' tests. I assumes the differences are following: 

 - functional tests are using multiple processes both: for servers and clients, while linearizable are combining (at least clients). 
 - functional tests were focusing whether after failure the cluster will recover, i.e. will be able to continue serving, not verifying the data correctness against requests. 
 - functional were more 'randomized' while linearizable are aiming at more determinsm. 

Are there other fundamental differences ?

Reply:
Based on the discussion in https://github.com/etcd-io/etcd/issues/14820, we need to unify the linearizablity and function test. Eventually we will only maintain one unified test.

Reply:
From discussion above, seems like contributors should not be adding any new functional tests but add new tests here. It would help if this Readme can clarify.